### PR TITLE
UW-547 Enable Python 3.12 support

### PR DESCRIPTION
**Pull request by maddenp-noaa**
**Synopsis**

- Enable Python 3.12 support.
- Update run and test packages.
- Update some tests.
- Change calls to deprecated `utcnow()` function.

**Type**

- [x] Code maintenance (refactoring, etc. without behavior change)

**Impact**

- [x] This is a non-breaking change (existing functionality continues to work as expected)

**Checklist**

- [x] I have added myself and any co-authors to the PR's _Assignees_ list.
- [x] I have reviewed the documentation and have made any updates necessitated by this change.

### CODE REVIEW

File: recipe/build.sh [hunk 1]
```diff
@@ -6,4 +6,5 @@ dst=../test_files
 rm -fr $dst
 mkdir -pv $dst
 mv -v $PKG_NAME $dst/
+ln -s $dst/$PKG_NAME
```
Comment (maddenp-noaa) [inline recipe/build.sh:9]: Fixes a problem observed when running `make package`, related to the recent introduction of the `--version` switch and its use of `src/uwtools/resources/info.json`.

File: src/uwtools/tests/api/test_chgres_cube.py [hunk 2]
```diff
@@ -13,7 +13,7 @@ def test_execute(tmp_path):
     args: dict = {
         "batch": False,
         "config": "config.yaml",
-        "cycle": dt.datetime.utcnow(),
+        "cycle": dt.datetime.now(),
```
Comment (maddenp-noaa) [inline src/uwtools/tests/api/test_chgres_cube.py:16]: `utcnow()` is deprecated and `pytest` warned about it.

File: src/uwtools/tests/test_cli.py [hunk 3]
```diff
@@ -696,7 +696,7 @@ def test_main_fail_exception_log():
                 with raises(SystemExit) as e:
                     cli.main()
                 assert e.value.code == 1
-            assert log.called_once_with(msg)
+            log.error.assert_called_once_with(msg)
```
Comment (maddenp-noaa) [inline src/uwtools/tests/test_cli.py:699]: Bug fixed -- the old `assert` was not doing what was intended. This is the correct syntax.

File: recipe/meta.json [hunk 4]
```diff
@@ -28,7 +28,7 @@
       "jinja2 =3.1.*",
       "jsonschema =4.20.*",
       "lxml =4.9.*",
-      "python >=3.9,<3.12",
+      "python >=3.9,<3.13",
       "pyyaml =6.0.*"
     ]
   },
```
Comment (maddenp-noaa) [inline recipe/meta.json:34]: Auto-generated by `make meta` based on contents of `meta.yaml`.

File: recipe/meta.yaml [hunk 5]
```diff
@@ -14,22 +14,22 @@ requirements:
...
     - isort 5.13.*
     - jq 1.7.*
     - make >=3.8
-    - mypy 1.8.*
+    - mypy 1.9.*
     - pylint 3.1.*
-    - pytest 7.4.*
+    - pytest 8.1.*
 about:
   license: LGPL
```
Comment (maddenp-noaa) [inline recipe/meta.yaml:35]: I took the opportunity to update and re-pin our run and test packages.

File: src/uwtools/api/chgres_cube.py [hunk 6]
```diff
@@ -1,6 +1,7 @@
 """
 API access to the uwtools ``chgres_cube`` driver.
 """
+
```
Comment (maddenp-noaa) [inline src/uwtools/api/chgres_cube.py:4]: The new `black` version added some newlines between module docstrings and code.

File: src/uwtools/config/jinja2.py [hunk 7]
```diff
@@ -46,11 +46,15 @@ def __init__(
...
-                else []
+                searchpath=(
+                    searchpath
+                    if searchpath
+                    else (
+                        self._template_source.parent
+                        if isinstance(self._template_source, Path)
+                        else []
+                    )
+                )
```
Comment (maddenp-noaa) [inline src/uwtools/config/jinja2.py:57]: New `black` behavior.

File: src/uwtools/tests/test_schemas.py [hunk 8]
```diff
@@ -179,7 +179,7 @@ def test_schema_files_to_stage():
     # A str -> str dict is ok:
     assert not errors({"file1": "/path/to/file1", "file2": "/path/to/file2"})
     # An empty dict is not allowed:
-    assert "does not have enough properties" in errors({})
+    assert "{} should be non-empty" in errors({})
```
Comment (maddenp-noaa) [inline src/uwtools/tests/test_schemas.py:182]: The new `jsonschema` has an updated error message for this case.

**Review by christinaholtNOAA — APPROVED** [review]
LGTM.

**Review by elcarpenterNOAA — APPROVED** [review]
LGTM.

**Review by NaureenBharwaniNOAA — APPROVED** [review]

**Comment by WeirAE** [comment]
Interesting that we're already seeing changes and depreciations in the new version! I think we're starting to get quite a bit ahead of the other apps, so we should consider talking them in to updating and/or move the functional tests up to ensure we don't break anything

**Review by WeirAE — APPROVED** [review]
LGTM

**Pull request merged**